pyomo/solvers/plugins/solvers/SAS.py

```diff
@@ -0,0 +1,794 @@
+            results.solver.termination_condition = TerminationCondition.userInterrupt
+            if solution_status != "ERROR":
+                results.solver.message = results.solver.termination_message = (
+                    SOLSTATUS_TO_MESSAGE[solution_status]
+                )
+        else:
+            results.solver.termination_condition = TerminationCondition.error
+            results.solver.message = results.solver.termination_message = (
+                SOLSTATUS_TO_MESSAGE["FAILED"]
+            )
+        return results
+
+    @abstractmethod
+    def _apply_solver(self):
+        pass
+
+    def _postsolve(self):
+        """Clean up at the end, especially the temp files."""
+        # Let the base class deal with returning results.
+        results = super(SASAbc, self)._postsolve()
+
+        # Finally, clean any temporary files registered with the temp file
+        # manager, created populated *directly* by this plugin. does not
+        # include, for example, the execution script. but does include
+        # the warm-start file.
+        TempfileManager.pop(remove=not self._keepfiles)
+
+        return results
+
+    def warm_start_capable(self):
+        """True if the solver interface supports MILP warmstarting."""
+        return True
+
+
+@SolverFactory.register("_sas94", doc="SAS 9.4 interface")
+class SAS94(SASAbc):
+    """
+    Solver interface for SAS 9.4 using saspy. See the saspy documentation about
+    how to create a connection.
+    """
+
+    def __init__(self, **kwds):
+        """Initialize the solver interface and see if the saspy package is available."""
+        super(SAS94, self).__init__(**kwds)
+
+        try:
+            import saspy
+
+            self._sas = saspy
+        except ImportError:
+            self._python_api_exists = False
+        except Exception as e:
+            self._python_api_exists = False
+            # For other exceptions, raise it so that it does not get lost
+            raise e
+        else:
+            self._python_api_exists = True
+            self._sas.logger.setLevel(logger.level)
+
+        # Create the session only as its needed
+        self._sas_session = None
+
+        # Store other options for the SAS session
+        self._session_options = kwds
+
+    def __del__(self):
+        # Close the session, if we created one
+        if self._sas_session:
+            self._sas_session.endsas()
+
+    def _create_statement_str(self, statement):
+        """Helper function to create the strings for the statements of the proc OPTLP/OPTMILP code."""
+        stmt = self.options.pop(statement, None)
+        if stmt:
+            return (
+                statement.strip()
+                + " "
+                + " ".join(option + "=" + str(value) for option, value in stmt.items())
+                + ";"
+            )
+        else:
+            return ""
+
+    def sas_version(self):
+        return self._sasver
+
+    def _apply_solver(self):
```

Comment:
This is a really big function. I recommend breaking it up into some smaller functions.